pkg/coscheduling/coscheduling.go

```diff
@@ -0,0 +1,257 @@
+	if podGroupName == "" || minAvailable <= 1 {
+		return framework.NewStatus(framework.Success, ""), 0
+	}
+
+	running := cs.calculateRunningPods(podGroupName)
+	waiting := cs.calculateWaitingPods(podGroupName)
+
+	if running+waiting+1 < minAvailable {
+		klog.V(3).Infof("The count of podGroup %v is not up to minAvailable(%d) in Permit: running(%d), waiting(%d)",
+			podGroupName, minAvailable, running, waiting)
+		return framework.NewStatus(framework.Wait, ""), 10 * time.Second
+	}
+
+	klog.V(3).Infof("The count of podGroup %v is up to minAvailable(%d) in Permit: running(%d), waiting(%d)",
+		podGroupName, minAvailable, running, waiting)
+	cs.FrameworkHandle.IterateOverWaitingPods(func(p framework.WaitingPod) {
+		if p.GetPod().Labels[PodGroupName] == podGroupName {
+			klog.V(3).Infof("Permit allows the pod: %v/%v", podGroupName, p.GetPod().Name)
+			p.Allow(cs.Name())
+		}
+	})
+
+	delete(cs.PodGroupInfos, podGroupName)
+	return framework.NewStatus(framework.Success, ""), 0
+}
+
+// Unreserve is called by the scheduling framework when a reserved pod was
+// rejected in a later phase.
+func (cs *Coscheduling) Unreserve(ctx context.Context, state *framework.CycleState, p *v1.Pod, nodeName string) {
+	podGroupName, exist := p.Labels[PodGroupName]
+	if !exist {
+		return
+	}
+
+	cs.FrameworkHandle.IterateOverWaitingPods(func(p framework.WaitingPod) {
+		if p.GetPod().Labels[PodGroupName] == podGroupName {
+			klog.V(3).Infof("Unreserve rejects the pod: %v/%v", podGroupName, p.GetPod().Name)
+			p.Reject(cs.Name())
+		}
+	})
+}
+
+func (cs *Coscheduling) checkPod(p *v1.Pod) (string, int) {
+	podGroupName, exist := p.Labels[PodGroupName]
+	if !exist {
+		return "", 0
+	}
+	minAvailable, exist := p.Labels[PodGroupMinAvailable]
+	if !exist {
+		return "", 0
+	}
+	minNum, err := strconv.Atoi(minAvailable)
+	if err != nil {
+		klog.Error(err)
+	}
+	klog.V(3).Infof("Pod is in a podGroup: %v/%v, minAvailable is %v", podGroupName, p.Name, minNum)
+	return podGroupName, minNum
+}
+
+func (cs *Coscheduling) calculateTotalPods(podGroupName string) int {
+	// Returns the list of pods by the podGroupName.
+	selector := labels.Set(labels.Set{PodGroupName: podGroupName}).AsSelector()
+	pods, err := cs.PodLister.List(selector)
+	if err != nil {
+		klog.Error(err)
+	}
+	return len(pods)
+}
+
+func (cs *Coscheduling) calculateAllocatedPods(podGroupName string) int {
+	// Returns the list of pods by the podGroupName.
+	selector := labels.Set(labels.Set{PodGroupName: podGroupName}).AsSelector()
+	pods, err := cs.FrameworkHandle.SnapshotSharedLister().Pods().List(selector)
+	if err != nil {
+		klog.Error(err)
+		return 0
+	}
+	return len(pods)
+}
+
+func (cs *Coscheduling) calculateRunningPods(podGroupName string) int {
+	// Returns the list of pods by the podGroupName.
+	selector := labels.Set(labels.Set{PodGroupName: podGroupName}).AsSelector()
+	pods, err := cs.FrameworkHandle.SnapshotSharedLister().Pods().List(selector)
```

Comment:
ditto

Reply:
And the pod status can be filtered at `FilteredList`.

Reply:
Tks, I change to filter the status by `FilteredList `